hnn_core/optimization/general_optimization.py

```diff
@@ -0,0 +1,407 @@
+"""Parameter optimization."""
+
+# Authors: Carolina Fernandez <cxf418@miami.edu>
+#          Nick Tolley <nicholas_tolley@brown.edu>
+#          Ryan Thorpe <ryan_thorpe@brown.edu>
+#          Mainak Jas <mjas@mgh.harvard.edu>
+
+import numpy as np
+
+from .metrics import _rmse_evoked
+
+from skopt import gp_minimize
+from scipy.optimize import fmin_cobyla
+
+
+class Optimizer:
+    def __init__(self, net, tstop, constraints, set_params, solver='bayesian',
+                 obj_fun='dipole_rmse', scale_factor=1.,
+                 smooth_window_len=None, max_iter=200):
+        """Parameter optimization.
+
+        Parameters
+        ----------
+        net : Network
+            The network object.
+        tstop : float
+            The simulated dipole's duration.
+        constraints : dict
+            The user-defined constraints.
+        set_params : func
+            User-defined function that sets parameters in network drives.
+        solver : string
+            The optimizer, 'bayesian' or 'cobyla'.
+        obj_fun : string
+            The objective function to be minimized.
+        scale_factor : float, optional
+            The dipole scale factor. The default is 1.
+        smooth_window_len : float, optional
+            The smooth window length. The default is None.
+        max_iter : int, optional
+            The max number of calls to the objective function. The default is
+            200.
+
+        Attributes
+        ----------
+        net : Network
+            The network object.
```

Comment:
I'm confused what is the difference between this `net` object and `net_`. I can imagine a user would be too. Should this `net` be provided to the init or is it better to have it in the `fit` method? Or maybe it should be a private attribute? Maybe you can get some inspiration from `sklearn` ?

Reply:
Makes sense. `net` is the network object the user passes in (has no drives). `net_` is an attribute the user can access after `fit` is called (it is a network objects with the final optimized drives). I think it makes sense then to make `net` private?

Reply:
private yes, and also worth thinking whether `net` and `tstop` should be arguments of the class init or in the `fit` method. Look around some libraries to get inspiration for API ... scipy, skopt, sklearn, keras etc

Reply:
I would even call it `initial_net`, `cortical_net` or something to that effect to avoid confusion for developers